### PR TITLE
Implement a basic bump-pointer allocator

### PR DESCRIPTION
**Pull request by jacob-hughes**
This is the very first stage of allocation. I've made this PR as small as possible because there's quite a lot of fiddly details coming up. This is a basic allocator which simply increments the pointer by the size of the last stored object until the heap runs out of memory.

### CODE REVIEW

File: src/collector.rs [hunk 1]
```diff
@@ -58,20 +55,84 @@ impl Collector {
...
         unimplemented!()
     }
 
-    // Perform the actual garbage collection. We use the name `reclaim` to
-    // disambiguate from Rust's notion of `collect` on iterators.
+    /// Perform the actual garbage collection. We use the name `reclaim` instead
+    /// of collect to disambiguate from Rust's notion of `collect` on iterators.
     pub(crate) fn reclaim(&self) {
-        unimplemented!()
+        eprintln!("Collection is no-op: not yet implemented")
```
Comment (ltratt) [inline src/collector.rs:68]: `reclaim is a no-op`?

**Comment by ltratt** [comment]
I'm not sure a bump-pointer allocator is a good idea given we're doing a non-moving collector at first. Wouldn't it be easier to just use malloc's malloc/free, since most of the complexity at first is going to be in scanning the stack with LLVM, finding the right places to put stuff in the Rust compiler etc? We can come back later and change to a bump pointer alloc IMHO.

**Comment by vext01** [comment]
(holding off my review until Laurie's comment is addressed)

**Comment by jacob-hughes** [comment]
> Wouldn't it be easier to just use malloc's malloc/free

I'm not quite sure what you mean. In a sense I already am doing this: the GC heap is a big pre-allocated chunk of memory which we then manage. I think `malloc` and `free`ing individual objects changes the way that you'll have to think about writing a mark-sweep collection algorithm. The way I have it now you can pretty much just lift it straight from the literature.

The alternative to bump pointer allocation for a mark-sweep collector is a free-list allocator which is more complex. I chose BP allocation not as an optimisation, but because it's the most straightforward way to allocate memory to a single heap.

File: src/collector.rs (hunk 1 above)
Comment (jacob-hughes) [reply src/collector.rs:68]: I mean this in the sense that at the moment, collection does nothing (But `unimplemented!` will panic, which is not desirable)

**Comment by ltratt** [comment]
Bump allocation makes most sense with a moving collector, otherwise we end up with particularly annoying fragmentation problems. Bump allocation is entirely orthogonal to issues like how we scan the stack etc. I'm suggesting something simpler than this PR is suggesting: if we just hand out a pointer we got from the system malloc then actual GCing memory simply means calling `free` (and we let the system malloc worry about fragmentation etc. etc.).

**Comment by jacob-hughes** [comment]
Ah okay I see, so you suggest not worrying about a managed heap at this point? In which case, I should probably close this PR and open a separate one for system `malloc`ing objects.

**Comment by ltratt** [comment]
That gets my vote!

**Pull request closed without merging**